google/cloud/spanner/client.cc

```diff
@@ -99,16 +99,68 @@ StatusOr<std::int64_t> Client::ExecutePartitionedDml(
   return Status(StatusCode::kUnimplemented, "not implemented");
 }
 
-StatusOr<CommitResult> Client::Commit(
-    Transaction const& /*transaction*/,
-    std::vector<Mutation> const& /*mutations*/) {
-  return Status(StatusCode::kUnimplemented, "not implemented");
+StatusOr<CommitResult> Client::Commit(Transaction const& transaction,
+                                      std::vector<Mutation> const& mutations) {
+  auto session = GetSession();
+  if (!session) {
+    return std::move(session).status();
+  }
+  google::spanner::v1::CommitRequest request;
+  request.set_session(session->session_name());
+  for (auto const& m : mutations) {
+    *request.add_mutations() = m.as_proto();
+  }
+  std::function<StatusOr<CommitResult>(spanner_proto::TransactionSelector&)>
+      commit([&request, this](spanner_proto::TransactionSelector& s)
+                 -> StatusOr<CommitResult> {
+        if (!s.id().empty()) {
+          request.set_transaction_id(s.id());
+        } else if (s.has_begin()) {
+          *request.mutable_single_use_transaction() = s.begin();
+        } else if (s.has_single_use()) {
+          *request.mutable_single_use_transaction() = s.single_use();
+        }
+        grpc::ClientContext context;
+        auto response = stub_->Commit(context, request);
+        if (!response) {
+          return std::move(response).status();
+        }
+        CommitResult r;
+        r.commit_timestamp = internal::FromProto(response->commit_timestamp());
+        return r;
```

Comment:
> Can you take another look and see if I am using the Transaction class as you intended?

Yes.  A simplifying incantation might make `Client::Commit()` just a `Visit()` call forwarding to an implementation function.

```
StatusOr<CommitResult> Client::Commit(Transaction transaction,
                                      std::vector<Mutation> const& mutations) {
  return internal::Visit(std::move(transaction),
                         [this, mutations](spanner_proto::TransactionSelector& selector) {
                           return conn_->Commit(selector, mutations);
                         });
} 
```

I believe this will be more like the pattern for things like `Read()`, where there are three overloads forwarding to a single implementation, and where that implementation is a concrete sub-class of `Connection` (as proposed in #305).

`Commit()` is also a little easier than things like `Read()` because the latter can return a new transaction id that we need to stuff into `selector`, whereas `Commit()` never needs to do that (it doesn't even return a transaction id).

Reply:
Ack. The incantation as-written did not work for me (see https://godbolt.org/z/WFB0MY), so I am explicitly casting the lambda to the right `std::function` for now.